### PR TITLE
[SPARK-23367][Build] Include python document style checking 

### PR DESCRIPTION
**Pull request by rekhajoshm**
## What changes were proposed in this pull request?
Includes python document style checking.
- Use sphinx like check, run only if pydocstyle installed on machine/jenkins
- use pydocstyle rather than single file version pep257.py, which  is much older and had some known issues
- verify pydocstyle latest 3.0.0  is in use, to ensure latest doc checks are getting executed
- ignore (inclusion/exclusion error codes) features and support via tox.ini
- Be non-breaking change and allow updating docstyle to standards at easy pace

## How was this patch tested?
./dev/run-tests

### CODE REVIEW

File: dev/tox.ini [hunk 1]
```diff
@@ -14,6 +14,8 @@
 # limitations under the License.
 
 [pycodestyle]
-ignore=E226,E241,E305,E402,E722,E731,E741,W503,W504
+ignore=E226,E241,E305,E402,E722,E731,E741,W503,W504,W605
```
Comment (srowen) [inline dev/tox.ini]: Do we need to exclude 605? I thought we fixed all those...

File: dev/tox.ini [hunk 2]
```diff
@@ -14,6 +14,8 @@
...
 max-line-length=100
 exclude=cloudpickle.py,heapq3.py,shared.py,python/docs/conf.py,work/*/*.py,python/.eggs/*,dist/*
+[pydocstyle]
+ignore=D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D203,D204,D205,D206,D207,D208,D209,D210,D211,D212,D213,D214,D215,D300,D301,D302,D400,D401,D402,D403,D404,D405,D406,D407,D408,D409,D410,D411,D412,D413,D414
```
Comment (srowen) [inline dev/tox.ini]: This is tiny, but is this blank line necessary? it seems to separate the heading from its content?

**Review by srowen — COMMENTED** [review]
I like the general idea, but this also turns off most checks? are there any that can be meaningfully re-enabled with a few fixes?

**Review by HyukjinKwon — COMMENTED** [review]
I would also leave short explanations in PR description to justify why they should be skipped. Looks so many things are being skipped. FYI, this project does not yet claim PEP 257 as far as I know.

File: dev/tox.ini (hunk 2 above)
Comment (rekhajoshm) [reply dev/tox.ini]: Between [pydocstyle] to ignore is necessary and standard style. Post ignore the blank line was  added as per requested by last reviewer on previous PR - https://github.com/apache/spark/pull/20556#discussion_r168079061

File: dev/tox.ini (hunk 1 above)
Comment (rekhajoshm) [reply dev/tox.ini]: The intent of this PR was on pydocstyle changes, and this was pycodestyle break, so was unsure if that fix can be included in this PR. I had seen W605 code style broken in that build, and to make it pass added this. I agree completely the fix for W605 can be done with no-PR trivial fix., will check again.

**Comment by rekhajoshm** [comment]
Sorry, I was waiting from Feb2018 to Jul2018 on the previous PR  #20556  , and hence there is a bit of disconnect from this issue and delay in providing context in which this PR was created.

As per earlier discussions with prior reviewers, the intent of this PR was to make changes to enable pydocstyle but not to include all pydocstyle changes across all docs as that would need to be discussed later and be prioritized. I will relook at this again with your comments and check pydoc for style corrections.

**Review by srowen — COMMENTED** [review]
Ok I think it's fine to focus on enabling this tool here and then making changes later as we unexclude warnings. 

I think we had separately fixed the w605 errors; might be worth checking if that's still needed.

File: dev/tox.ini (hunk 2 above)
Comment (holdenk) [reply dev/tox.ini]: I don't think that's what @ueshin was asking for, I think it was a blank line after the `ignore=...`, but if @ueshin is around we can see what @ueshin says. It's also relatively minor provided everything functions.

File: dev/tox.ini (hunk 1 above)
Comment (holdenk) [reply dev/tox.ini]: I'm just confused why this would need to be changed in this PR -- hopefully just a hold over from the previous PR?

**Review by holdenk — COMMENTED** [review]
Thanks for reviving this PR I'm excited to get better style checkers integrated into Spark. My main concern is with the change to pycodestyle's list of excluded problems.

File: dev/tox.ini (hunk 2 above)
Comment (ueshin) [reply dev/tox.ini]: I just asked to add a line break at the end of file, and the current style looks good to me.

**Comment by holdenk** [comment]
Gentle ping, whats up?

**Comment by rekhajoshm** [comment]
Thank you @srowen @holdenk @HyukjinKwon for your review.

As per discussion, in this PR, we are enabling docstyle change but not updating pydocs. Hence tox.ini has doc error codes included.It can be taken off few at a time.

pydocstyle is newer correct way of handling doc style issues over pep257. It covers both pep257 and numpy convention style.  http://www.pydocstyle.org/en/3.0.0/error_codes.html

updated pydocstyle to latest 3.0.0.
removed W605 from tox.ini. (It had appeared in the build when this branch was created, not lingering from old PR.)

File: dev/lint-python [hunk 3]
```diff
@@ -99,6 +104,29 @@ else
     echo "flake8 checks passed."
 fi
 
+# Check python document style, skip check if pydocstyle is not installed.
```
Comment (srowen) [inline dev/lint-python]: Nit: the indentation below looks different (4-space) from the rest of the script (2-space)

File: dev/tox.ini (hunk 2 above)
Comment (srowen) [reply dev/tox.ini]: Oh I was asking about the blank line at line 21. Isn't "ignore=" related to "[pydocstyle]"? compare to no blank line between 16-17. It's really trivial but hey.

File: dev/lint-python (hunk 3 above)
Comment (srowen) [reply dev/lint-python]: Actually, seems like indentation is inconsistent in this file, so let's leave it for now

**Pull request closed without merging**